src/uu/fold/src/fold.rs

```diff
@@ -92,123 +91,173 @@ fn fold(filenames: Vec<String>, bytes: bool, spaces: bool, width: usize) {
-                                                    0
-                                                }
-                                            }
-                                            '\r' => return 0,
-                                            _ => 1,
-                                        }
-                                    });
-                                    (&slice[0..=m], routput, ncount)
-                                }
-                                None => (slice, "", 0),
-                            }
-                        } else {
-                            (slice, "", 0)
-                        };
-                        println!("{}", out);
-                        (val.to_owned(), ncount)
-                    };
-                    output = val;
-                    count = ncount;
-                }
-                match ch {
-                    '\t' => {
-                        count += 8;
-                        if count > width {
-                            println!("{}", output);
-                            output.truncate(0);
-                            count = 8;
-                        }
-                    }
-                    '\x08' => {
-                        if count > 0 {
-                            count -= 1;
-                            let len = output.len() - 1;
-                            output.truncate(len);
-                        }
-                        continue;
-                    }
-                    '\r' => {
-                        output.truncate(0);
-                        count = 0;
-                        continue;
-                    }
-                    _ => count += 1,
-                };
-                output.push(ch);
+            };
+
+            // Don't duplicate trailing newlines: if the slice is "\n", the
+            // previous iteration folded just before the end of the line and
+            // has already printed this newline.
+            if slice == "\n" {
+                break;
             }
-            if count > 0 {
-                println!("{}", output);
+
+            i += slice.len();
+
+            let at_eol = i >= len;
+
+            if at_eol {
+                print!("{}", slice);
+            } else {
+                println!("{}", slice);
             }
         }
+
+        line.truncate(0);
     }
 }
 
-#[inline]
-fn rfind_whitespace(slice: &str) -> Option<usize> {
-    for (i, ch) in slice.chars().rev().enumerate() {
-        if ch.is_whitespace() {
-            return Some(slice.chars().count() - (i + 1));
+/// Fold `file` to fit `width` (number of columns).
+///
+/// By default `fold` treats tab, backspace, and carriage return specially:
+/// tab characters count as 8 columns, backspace decreases the
+/// column count, and carriage return resets the column count to 0.
+///
+/// If `spaces` is `true`, attempt to break lines at whitespace boundaries.
+#[allow(unused_assignments)]
```

Comment:
Why did you add this directive? thanks

Reply:
Hi @sylvestre. Without this directive, the compiler complains about the match arm handling `\t` on lines 222–228:

```rust
                '\t' => {
￼                    if col_count + 8 > width && !output.is_empty() {
￼                        emit_output!();
￼                    }
￼                    col_count += 8;
￼                    last_space = Some(char_count);
                }
```
https://github.com/uutils/coreutils/pull/2025/files#diff-c0cfac636e7701f77f3693e680acd57c6c41eac32ccf3cb63ea30738ebce8687R222-R227

Here is the compiler output without the directive:

```
$ cargo build -p uu_fold
warning: value assigned to `last_space` is never read
   --> src/uu/fold/src/fold.rs:198:13
    |
198 |             last_space = None;
    |             ^^^^^^^^^^
...
223 |                         emit_output!();
    |                         --------------- in this macro invocation
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?
```

Generally, we want `emit_output!()` to reset the last seen whitespace to `None` since we've just printed it out:

```
    output = "...foo bar baz"
last_space =            ^

<call emit_output!(), prints "...foo bar ">

    output = "baz"
last_space = None
```

Tab presents an edge case: if the tab pushes the output over `width`, we want to emit the output, but the tab character is now the last seen whitespace, so we overwrite the value set by the macro and point `last_space` at the tab.

Taking a fresh look at this, I think I see how to parameterize the macro to eliminate the need for the directive without triggering the compiler warning. Before I do that, though, does the `emit_output!()` macro improve the readability of the implementation? Is it an appropriate solution here?